File: module5_ws/src/cpp_number_publisher/src/main_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int32.hpp"

class NumberPublisher : public rclcpp::Node
{
public:
    NumberPublisher() : Node("number_publisher"), counter_(0)
    {
        timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&NumberPublisher::number_publisher, this));
        publisher_ = this->create_publisher<example_interfaces::msg::Int32>("number", 10);
        RCLCPP_INFO(this->get_logger(), "Number publisher node has been started.");
    }
    void number_publisher()
    {
        auto msg = example_interfaces::msg::Int32();
        msg.data = counter_;
        publisher_->publish(msg);
        RCLCPP_INFO(this->get_logger(), "Publishing: %d", msg.data);
        counter_++;
    }
private:
    int counter_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<example_interfaces::msg::Int32>::SharedPtr publisher_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberPublisher>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: module5_ws/src/number_publisher/number_publisher/main_node.py
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int32

class NumberPublisher(Node):
    def __init__(self):
        super().__init__('number_publisher')
        self.publisher_ = self.create_publisher(Int32, 'number', 10)
        self.timer = self.create_timer(1.0, self.publish_number)
        self.number = 0
        self.get_logger().info('Number Publisher Node has been started.')

    def publish_number(self):
        msg = Int32()
        msg.data = self.number
        self.publisher_.publish(msg)
        self.get_logger().info(f'Publishing: {self.number}')
        self.number += 1

def main(args=None):
    rclpy.init(args=args)
    number_publisher = NumberPublisher()
    rclpy.spin(number_publisher)
    rclpy.shutdown()